mesonbuild/modules/simd.py

```diff
@@ -13,76 +13,119 @@
-            raise mesonlib.MesonException('Compiler argument must be a compiler object.')
         conf = build.ConfigurationData()
-        for iset in self.isets:
+
+        static_lib_kwargs = T.cast('T.Dict[str, T.Any]', kwargs.copy())
+        del static_lib_kwargs['compiler']
+        for iset in ISETS:
+            del static_lib_kwargs[iset]
+        static_lib_kwargs = T.cast('kwtypes.StaticLibrary', static_lib_kwargs) # type: ignore
+
+        for iset in ISETS:
             if iset not in kwargs:
                 continue
-            iset_fname = kwargs[iset] # Might also be an array or Files. static_library will validate.
-            args = compiler.get_instruction_set_args(iset)
-            if args is None:
-                mlog.log('Compiler supports %s:' % iset, mlog.red('NO'))
+
+            sources = kwargs.get(iset, [])
+            if not sources:
                 continue
-            if args:
-                if not compiler.has_multi_arguments(args, state.environment)[0]:
-                    mlog.log('Compiler supports %s:' % iset, mlog.red('NO'))
-                    continue
-            mlog.log('Compiler supports %s:' % iset, mlog.green('YES'))
-            conf.values['HAVE_' + iset.upper()] = ('1', 'Compiler supports %s.' % iset)
-            libname = prefix + '_' + iset
-            lib_kwargs = {'sources': iset_fname,
-                          }
-            lib_kwargs.update(basic_kwargs)
-            langarg_key = compiler.get_language() + '_args'
-            old_lang_args = mesonlib.extract_as_list(lib_kwargs, langarg_key)
-            all_lang_args = old_lang_args + args
-            lib_kwargs[langarg_key] = all_lang_args
-            result.append(self.interpreter.func_static_lib(None, [libname], lib_kwargs))
-        return [result, conf]
-
-def initialize(*args, **kwargs):
+
+            cargs = compiler.get_instruction_set_args(iset)
+            if cargs is None:
+                mlog.log(f'Compiler supports {iset}:', mlog.red('NO'))
+                continue
+
+            if not compiler.has_multi_arguments(cargs, state.environment)[0]:
+                mlog.log(f'Compiler supports {iset}:', mlog.red('NO'))
+                continue
+
+            mlog.log(f'Compiler supports {iset}:', mlog.green('YES'))
+            conf.values['HAVE_' + iset.upper()] = ('1', f'Compiler supports {iset}.')
+
+            my_name = f'{prefix}_{iset}'
+
+            my_kwargs = static_lib_kwargs.copy()
+            my_kwargs['sources'] = sources
+
+            # Add compile args we derived above to those the user provided us
+            lang_args_key = compiler.get_language() + '_args'
+            old_lang_args = mesonlib.extract_as_list(my_kwargs, lang_args_key)
+            all_lang_args = old_lang_args + cargs
+            my_kwargs[lang_args_key] = all_lang_args
+
+            # TODO: Seems like a layering violation
+            lib = self.interpreter.func_static_lib(state.current_node, [my_name], my_kwargs)
```

Comment:
I had looked at this before, I think it makes more sense to use:
```python
lib = self.interpreter.build_target(state.current_node, (my_name, ), my_kwargs, build.StaticLibrary)
```
which avoids running back through the validators.